Type mismatch helper not check is closure return block

### Code

```Rust
struct SmolStr;
const _: fn() = || {
    match Some(()) {
        Some(()) => (),
        None => return,
    };
    let _: String = {
        SmolStr
    };
};
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:8:9
  |
8 |         SmolStr
  |         ^^^^^^^ expected `String`, found `SmolStr`
  |
note: return type inferred to be `String` here
 --> src/lib.rs:5:17
  |
5 |         None => return,
  |                 ^^^^^^
```

### Desired output

```Shell
error[E0308]: mismatched types
 --> src/lib.rs:8:9
  |
8 |         SmolStr
  |         ^^^^^^^ expected `String`, found `SmolStr`
  |
```

### Rationale and extra context

_No response_

### Other cases

```Rust
struct SmolStr;
const _: fn() = || {
    match Some(()) {
        Some(()) => (),
        None => return,
    };
    let _: String = SmolStr;
};
```

### Rust Version

```Shell
rustc 1.97.0-nightly (913e4bea8 2026-04-22)
binary: rustc
commit-hash: 913e4bea83424658d76712fee9c52452a3a9ef0e
commit-date: 2026-04-22
host: aarch64-unknown-linux-gnu
release: 1.97.0-nightly
LLVM version: 22.1.2
```

### Anything else?

_No response_

Regressed nightly: 2022-09-17,
nightly-2022-09-16 output: 
```
error[E0308]: mismatched types
 --> /data/exp/repos/rust/tmp-closure-type-mismatch.rs:8:9
  |
8 |         SmolStr
  |         ^^^^^^^ expected struct `String`, found struct `SmolStr`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

Seems to point to #101753

--- a/compiler/rustc_hir_typeck/src/coercion.rs
+++ b/compiler/rustc_hir_typeck/src/coercion.rs
@@ -2009,6 +2009,9 @@ impl<'tcx> CoerceMany<'tcx> {
             // note in this case, since it would be incorrect.
             && let Some(fn_sig) = fcx.body_fn_sig()
             && fn_sig.output().is_ty_var()
+            && fcx.ret_coercion.as_ref().is_some_and(|ret_coercion| {
+                fcx.resolve_vars_if_possible(ret_coercion.borrow().expected_ty()) == expected
+            })
         {
             err.span_note(sp, format!("return type inferred to be `{expected}` here"));
         }

--- a/tests/ui/closures/closure-return-block-note-issue-155670.rs
+++ b/tests/ui/closures/closure-return-block-note-issue-155670.rs
@@ -1,0 +1,14 @@
+struct SmolStr;
+
+const _: fn() = || {
+    match Some(()) {
+        Some(()) => (),
+        None => return,
+    };
+    let _: String = {
+        SmolStr
+        //~^ ERROR mismatched types
+    };
+};
+
+fn main() {}

--- a/tests/ui/closures/closure-return-block-note-issue-155670.stderr
+++ b/tests/ui/closures/closure-return-block-note-issue-155670.stderr
@@ -1,0 +1,9 @@
+error[E0308]: mismatched types
+  --> $DIR/closure-return-block-note-issue-155670.rs:9:9
+   |
+LL |         SmolStr
+   |         ^^^^^^^ expected `String`, found `SmolStr`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0308`.

